AccessTokenManagement updates (#793)
DuendeSoftware/docs.duendesoftware.com: src/content/docs/accesstokenmanagement/workers.mdx

---
title: Service Workers and Background Tasks
date: 2024-10-12
description: Learn how to manage OAuth access tokens in worker applications and background tasks using Duende.AccessTokenManagement.
sidebar:
  label: Service Workers
  order: 2
redirect_from:
  - /foss/accesstokenmanagement/workers/
---
import {  Badge } from "@astrojs/starlight/components";
import { Tabs, TabItem } from "@astrojs/starlight/components";

A common scenario in worker applications or background tasks (or really any daemon-style applications) is to call APIs using an OAuth token obtained via the client credentials flow.

The access tokens need to be requested and [cached](/accesstokenmanagement/advanced/client-credentials/#token-caching) (either locally or shared between multiple instances) and made available to the code calling the APIs. In case of expiration (or other token invalidation reasons), a new access token needs to be requested.

The actual business code should not need to be aware of this.

> Have a look for the [Worker project in the samples folder](https://github.com/DuendeSoftware/foss/tree/main/access-token-management/samples/) for running code.

Author: ```suggestion
Take a look at the [`Worker` project in the samples folder](https://github.com/DuendeSoftware/foss/tree/main/access-token-management/samples/) for example code.
```

> For more information, also please see the [advanced topic on client credentials](accesstokenmanagement/advanced/client-credentials/). 

Author: ```suggestion
For more information, see the [advanced topic on client credentials](accesstokenmanagement/advanced/client-credentials/). 
```

## Usage

First, you'll need to add `Duende.AccessTokenManagement` to your solution. 

> There are two fundamental ways to interact with token management:
> 1. **Automatic** <Badge text="recommended"/>: You request a http client from the IHTTPClientFactory. This http client automatically requests, optionally renews and attaches the access tokens on each request. 
> 2. **Manually**  <Badge text="advanced"/>: You request an access token, which you can then use to (for example) authenticate with services. You are responsible for attaching the access token to requests. 

Author: ```suggestion
Next, there are two fundamental ways to interact with token management:
1. **Automatic** <Badge text="recommended"/>: You request an `HttpClient` from the `IHttpClientFactory`. This HTTP client automatically requests, optionally renews and attaches the access tokens on each request. 
2. **Manually**  <Badge text="advanced"/>: You request an access token, which you can then use to (for example) authenticate with services. You are responsible for attaching the access token to requests. 
```

### Adding Duende.AccessTokenManagement

Start by adding a reference to the `Duende.AccessTokenManagement` NuGet package to your application.

```bash
dotnet add package Duende.AccessTokenManagement
```

You can add the necessary services to the ASP.NET Core service provider by calling `AddClientCredentialsTokenManagement()`. After that you can add one or more named client definitions by calling `AddClient`.

> :::note
>     The way you interact with `Duende.AccessTokenManagement` has changed in version 4. 
> :::
> 

Author: ```suggestion
```

I'd drop this note tbh. The code sample shows different approaches.

        });
    ```
  </TabItem>
  <TabItem label="V3">

    ```csharp
    // Program.cs
    services.AddClientCredentialsTokenManagement()
        .AddClient("catalog.client", client =>
        {
            client.TokenEndpoint = "https://demo.duendesoftware.com/connect/token";

            client.ClientId = "6f59b670-990f-4ef7-856f-0dd584ed1fac";
            client.ClientSecret = "d0c17c6a-ba47-4654-a874-f6d576cdf799";

            client.Scope = "catalog inventory";
        })
        .AddClient("invoice.client", client =>
        {
            client.TokenEndpoint = "https://demo.duendesoftware.com/connect/token";

            client.ClientId = "ff8ac57f-5ade-47f1-b8cd-4c2424672351";
            client.ClientSecret = "4dbbf8ec-d62a-4639-b0db-aa5357a0cf46";

            client.Scope = "invoice customers";
        });

    // in v3, you explicitly need to add a distributed cache implementation, such as in-memory
    services.AddDistributedMemoryCache();

    ```
    </TabItem>
</Tabs>    

### Automatic token management using HTTP Factory

You can register HTTP clients with the factory that will automatically use the above client definitions to request and use access tokens.

> The following code registers an HTTP client called `invoices` to automatically use the `invoice.client` definition:

Author: ```suggestion
The following code registers an `HttpClient` called `invoices` which automatically uses the `invoice.client` definition:
```

    ```csharp
    // Program.cs
    services.AddHttpClient<CatalogClient>(client =>
        {
            client.BaseAddress = new Uri("https://apis.company.com/catalog/");
        })
        .AddClientCredentialsTokenHandler("catalog.client");
    ```

    </TabItem>
</Tabs>

    Once you have set up HTTP clients in the HTTP factory, then no token related code is needed at all, e.g.:

    ```csharp
    // WorkerHttpClient.cs
    public class WorkerHttpClient : BackgroundService
    {
        private readonly ILogger<WorkerHttpClient> _logger;
        private readonly IHttpClientFactory _clientFactory;

        public WorkerHttpClient(ILogger<WorkerHttpClient> logger, IHttpClientFactory factory)
        {
            _logger = logger;
            _clientFactory = factory;
        }

        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
        {
            while (!stoppingToken.IsCancellationRequested)
            {
                var client = _clientFactory.CreateClient("invoices");
                var response = await client.GetAsync("test", stoppingToken);

                // rest omitted
            }
        }
    }

>     ```

Author: Please decrease the indent here with one level.

> **remark** The clients in the factory have a message handler attached to them that automatically re-tries the request in case of a `401` response code. The request get re-sent with a newly requested access token. If this still results in a `401`, the response is returned to the caller.

Author: ```suggestion
:::note
The clients in the HTTP client factory have a message handler attached to them that automatically retries the request in case of a `401` response code. The request get resent with a newly requested access token. If this still results in a `401`, the response is returned to the caller.
:::
```
